sdk/storage/Azure.Storage.Common/src/AsyncCollection.cs

```diff
@@ -0,0 +1,207 @@
+
+    }
+
+    /// <summary>
+    /// A single <see cref="Page{T}"/> of values from a request that may return
+    /// zero or more <see cref="Page{T}"/>s of values.
+    /// </summary>
+    /// <typeparam name="T">The type of values.</typeparam>
+    public readonly struct Page<T> : IEquatable<Page<T>>
+    {
+// TODO: Should this be Items instead of Values?
+// - Probably not with Response.Values
+// TODO: Should it be IEnumerable<T>?
+        #pragma warning disable CA1819 // Properties should not return arrays
+        /// <summary>
+        /// Gets the values in this <see cref="Page{T}"/>.
+        /// </summary>
+        public T[] Values { get; }
+        #pragma warning restore CA1819 // Properties should not return arrays
+
+        // TODO: Should this be object instead of string?
+        /// <summary>
+        /// Gets the continuation token used to request the next
+        /// <see cref="Page{T}"/>.  The continuation token may be null or
+        /// empty when there are no more pages.
+        /// </summary>
+        public string ContinuationToken { get; }
+
+        /// <summary>
+        /// The <see cref="Response"/> that provided this <see cref="Page{T}"/>.
+        /// </summary>
+        private readonly Response _response;
+
+        /// <summary>
+        /// Gets the <see cref="Response"/> that provided this
+        /// <see cref="Page{T}"/>.
+        /// </summary>
+        public Response GetRawResponse() => this._response;
+
+        /// <summary>
+        /// Creates a new <see cref="Page{T}"/>.
+        /// </summary>
+        /// <param name="values">
+        /// The values in this <see cref="Page{T}"/>.
+        /// </param>
+        /// <param name="continuationToken">
+        /// The continuation token used to request the next <see cref="Page{T}"/>.
+        /// </param>
+        /// <param name="response">
+        /// The <see cref="Response"/> that provided this <see cref="Page{T}"/>.
+        /// </param>
+        public Page(T[] values, string continuationToken, Response response)
+        {
+            this.Values = values;
+            this.ContinuationToken = continuationToken;
+            this._response = response;
+        }
+
+        /// <summary>
+        /// Creates a string representation of an <see cref="Page{T}"/>.
+        /// </summary>
+        /// <returns>
+        /// A string representation of an <see cref="Page{T}"/>.
+        /// </returns>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override string ToString() => base.ToString();
+
+        /// <summary>
+        /// Check if two <see cref="Page{T}"/> instances are equal.
+        /// </summary>
+        /// <param name="obj">The instance to compare to.</param>
+        /// <returns>True if they're equal, false otherwise.</returns>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override bool Equals(object obj) =>
```

Comment:
Do we have a use case for Equality comparison on this type?

Reply:
Not that I'm specifically aware of here

Reply:
Let's start by not having it then, especially considering the implementation that compares value collection by reference.

Reply:
I don't think we could define equality well over pages.  Over the entire sequence, yes, in which case we could use SequenceEquals.

Reply:
Removing and suppressing resulting warning.